// @nyaruka/temba-components/out-tsc/src/tabpane/TabPane.js
import { __decorate } from "tslib";
import { css, html } from 'lit';
import { property } from 'lit/decorators';
import { RapidElement } from '../RapidElement';
export class TabPane extends RapidElement {
    constructor() {
        super(...arguments);
        this.index = 0;
    }
    static get styles() {
        return css `
      :host {
        display: flex;
        flex-direction: column;
        min-height: 0;
      }

      .tabs {
        display: flex;
      }

      .tab {
        padding: 0.5em 1em;
        margin: 0em 0em;
        cursor: pointer;
        display: flex;
        border-radius: var(--curvature);
        border-bottom-right-radius: 0px;
        border-bottom-left-radius: 0px;
        border: 0px solid rgba(0, 0, 0, 0.45);
        color: var(--color-text-dark);
        --icon-color: var(--color-text-dark);
      }

      .tab temba-icon {
        margin-right: 0.4em;
      }

      .tab.selected {
        cursor: default;
        box-shadow: 2px 1px 3px 2px rgba(0, 0, 0, 0.07);
        background: #fff;
      }

      .pane {
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        background: #fff;
        border-radius: var(--curvature);
        box-shadow: 2px 5px 12px 2px rgba(0, 0, 0, 0.09),
          3px 3px 2px 1px rgba(0, 0, 0, 0.05);
        min-height: 0;
      }

      .pane.first {
        border-top-left-radius: 0px;
      }
    `;
    }
    handleTabClick(event) {
        this.index = parseInt(event.currentTarget.dataset.index);
        this.requestUpdate('index');
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('index')) {
            if (this.children.length > this.index) {
                for (let i = 0; i < this.children.length; i++) {
                    const tab = this.children[i];
                    tab.selected = i == this.index;
                    if (tab.selected) {
                        tab.style.display = 'flex';
                    }
                    else {
                        tab.style.display = 'none';
                    }
                }
            }
        }
    }
    render() {
        const tabs = [];
        for (const tab of this.children) {
            tabs.push(tab);
        }
        return html `
      <div class="tabs">
        ${tabs.map((tab, index) => html `
            <div
              @click=${this.handleTabClick}
              data-index=${index}
              class="tab ${index == this.index ? 'selected' : ''}"
            >
              ${tab.icon ? html `<temba-icon name=${tab.icon} />` : null}
              ${tab.name}
            </div>
          `)}
      </div>
      <div class="pane ${this.index === 0 ? 'first' : null}">
        <slot></slot>
      </div>
    `;
    }
}
__decorate([
    property({ type: Number })
], TabPane.prototype, "index", void 0);
